### PR TITLE
CDAP-14738 ablity to use py4j library to run python scripts

### PR DESCRIPTION
**Pull request by aonischuk**
This patch adds classes to run python code with py4j and jython, which is configurable by widget.

Work remaining:
* implement mutual authentication for py4j
* write tests

### CODE REVIEW

File: docs/PythonEvaluator-transform.md [hunk 1]
```diff
@@ -21,6 +21,21 @@ For example:
 
 will scale the ``'count'`` field of ``record`` by 1024.
 
+**executionMode:**
```
Comment (albertshau) [inline docs/PythonEvaluator-transform.md]: new documentation uses the same label that appears in the UI. Can you change all the configuration names (including the pre-existing ones) to be like that.

Comment (albertshau) [reply docs/PythonEvaluator-transform.md]: Execution Model: How to execute the python code. Supported modes are:

File: docs/PythonEvaluator-transform.md [hunk 2]
```diff
@@ -21,6 +21,21 @@ For example:
...
+_Interpreted mode_:
+Python code is executed via jvm, hence C based libs (e.g. numpy) and Python3 syntax are not supported.
+
+_Native mode:_
+Python process is started, which removes limitations. However, user is required to install py4j library on all hosts.
```
Comment (albertshau) [inline docs/PythonEvaluator-transform.md]: Python is run natively, so any python version and libraries installed on executors can be used. Native mode requires that python is available on every node in your cluster. It also requires that the py4j library is available on every node. To install py4j with Python 2, run pip install py4j. To install py4j with Python 3, run pip3 install py4j.

File: docs/PythonEvaluator-transform.md [hunk 3]
```diff
@@ -21,6 +21,21 @@ For example:
...
+For Python 2 run ``pip install py4j`` to install it. For Python 3 run ``pip3 install py4j``.
+
+**pythonBinary:** Path to binary which will run the python code. E.g. ``/usr/bin/python3``.
+Works only in native mode.
```
Comment (albertshau) [inline docs/PythonEvaluator-transform.md]: This is only required in native mode.

File: docs/PythonEvaluator-transform.md [hunk 4]
```diff
@@ -21,6 +21,21 @@ For example:
...
+
+**pythonPath:** PYTHONPATH environment variable. Allows to include libs from various locations. Path are separated with a colon.
```
Comment (albertshau) [inline docs/PythonEvaluator-transform.md]: PYTHONPATH environment variable. This is used to include libraries from various locations on the executor nodes.

Comment (albertshau) [reply docs/PythonEvaluator-transform.md]: Remove the part about colon separator. We now only describe things that UI users can see.

File: src/main/java/co/cask/hydrator/python/transform/JythonPythonExecutor.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright © 2019
+ * Cask Data, Inc.
```
Comment (albertshau) [inline src/main/java/co/cask/hydrator/python/transform/JythonPythonExecutor.java]: should be on previous line

File: pom.xml [hunk 6]
```diff
@@ -241,6 +260,8 @@
           <instructions>
             <_exportcontents>
               co.cask.hydrator.python.transform.*;
+              co.cask.hydrator.common.*;
```
Comment (albertshau) [inline pom.xml:263]: how come these were needed?

File: src/main/java/co/cask/hydrator/python/transform/PythonEvaluator.java [hunk 7]
```diff
@@ -102,9 +89,42 @@
     @Nullable
     private final String schema;
 
+    @Description("In interpreted mode Python code is executed via jvm, hence C based libs (e.g. numpy) and " +
+      "Python3 syntax are not supported.\n" +
+      "In native mode external Python process is started, which removes limitations. However, user " +
+      "is required to install py4j library on all hosts. For Python 2 run \"pip install py4j\" to install it. " +
+      "For Python 3 run \"pip3 install py4j\".")
+    private String executionMode;
```
Comment (albertshau) [inline src/main/java/co/cask/hydrator/python/transform/PythonEvaluator.java:99]: these properties can be macro enabled (annotate with `@Macro`)

File: src/main/java/co/cask/hydrator/python/transform/Py4jPythonExecutor.java [hunk 8]
```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright © 2019
+ * Cask Data, Inc.
```
Comment (albertshau) [inline src/main/java/co/cask/hydrator/python/transform/Py4jPythonExecutor.java]: should be on previous line

File: src/main/java/co/cask/hydrator/python/transform/Py4jPythonExecutor.java [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+import java.io.PrintWriter;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
```
Comment (albertshau) [inline src/main/java/co/cask/hydrator/python/transform/Py4jPythonExecutor.java]: style nit: in our project we don't use static imports. Instead, use TimeUnit.SECONDS in the code and import java.util.concurrent.TimeUnit.

File: src/main/java/co/cask/hydrator/python/transform/Py4jPythonExecutor.java [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+  private static final int PYTHON_PROCESS_PORT_FILE_CREATED_INTERVAL = 1;
+  private static final int PYTHON_PROCESS_PORT_FILE_CREATED_TIMEOUT = 120;
+  private static final int PYTHON_PROCESS_STOP_TIMEOUT = 30;
+  private static final int PYTHON_PROCESS_KILL_TIMEOUT = 10;
+  private static final String USER_CODE_PLACEHOLDER = "\\$\\{cdap\\.transform\\.function\\}";
+  private static final String DEFAULT_PYTHON_BINARY = "python";
+  private static final String PYTHONPATH_ENV_VARIABLE_NAME = "PYTHONPATH";
+
+  private final PythonEvaluator.Config config;
+  private String pythonPath;
```
Comment (albertshau) [inline src/main/java/co/cask/hydrator/python/transform/Py4jPythonExecutor.java]: use final variables when possible. It's easier to reason about things that can never change. pythonPath and pythonBinary can be made final

File: src/main/java/co/cask/hydrator/python/transform/Py4jPythonExecutor.java [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    if (gatewayServer != null) {
+      gatewayServer.shutdown();
+    }
+
+    logger.debug("Waiting for python process to finish");
+
+    if (process == null) {
+      logger.warn("Process was not created. Not waiting for it to finish.");
```
Comment (albertshau) [inline src/main/java/co/cask/hydrator/python/transform/Py4jPythonExecutor.java]: don't think we need to log this, it should be clear that something already went wrong in the logs.

File: src/main/java/co/cask/hydrator/python/transform/Py4jPythonExecutor.java [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+      }
+    } catch (Exception e) {
+      logger.warn("Cannot close python process.\n", e);
+    }
+
+    if (gatewayServer != null) {
+      gatewayServer.shutdown();
+    }
+
+    logger.debug("Waiting for python process to finish");
```
Comment (albertshau) [inline src/main/java/co/cask/hydrator/python/transform/Py4jPythonExecutor.java]: move this after the process null check

File: src/main/java/co/cask/hydrator/python/transform/Py4jPythonExecutor.java [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+  private static final int PYTHON_PROCESS_KILL_TIMEOUT = 10;
+  private static final String USER_CODE_PLACEHOLDER = "\\$\\{cdap\\.transform\\.function\\}";
+  private static final String DEFAULT_PYTHON_BINARY = "python";
+  private static final String PYTHONPATH_ENV_VARIABLE_NAME = "PYTHONPATH";
+
+  private final PythonEvaluator.Config config;
+  private String pythonPath;
+  private String pythonBinary;
+  private File pythonProcessOutputFile;
+  private Logger logger;
```
Comment (albertshau) [inline src/main/java/co/cask/hydrator/python/transform/Py4jPythonExecutor.java]: private static final Logger LOGGER = LoggerFactory.getLogger(Py4jPythonExecutor.class)

**Review by albertshau — COMMENTED** [review]
Looks mostly good, just a few small comments.

One larger question I have is about what happens if the java process dies without getting a chance to call finish() on the transport. This might happen in certain scenarios where the program is forcibly killed, and it actually always happens when running with the Spark engine because the framework isn't able to call destroy() in spark pipeline.

It seems like we need some sort of mechanism that will allow the python process to know that the java process is no longer running and shut itself down. This should be done in a separate PR after we update the design doc.

Just off the top of my head, one option could be for the java side to start a GatewayServer first that the python connects to. Assuming the JavaGateway in python has some way to tell that it can no longer connect, it could use that to decide if it needs to shut itself down.

File: docs/PythonEvaluator-transform.md (hunk 3 above)
Comment (aonischuk) [reply docs/PythonEvaluator-transform.md]: It's not required actually in native mode. If you leave it empty than "python" from path will be used

File: pom.xml (hunk 6 above)
Comment (aonischuk) [reply pom.xml:263]: JavaScript transform (which is in core-plugins) and Python transform have some common classes I moved them to hydrator-commons in hydrator-plugins. (need to merge it though). This was actually described in previous PR (#1), however seems like I just didn't add these lines there for some reason.

Without this:
"java.lang.Exception: java.lang.NoClassDefFoundError: Class not found in all delegated ClassLoaders: co.cask.hydrator.common.script.ScriptContext"

**Comment by aonischuk** [comment]
@albertshau I have fixed all the issues you mentioned. Please feel free to check/merge.

Regrading "destroy" problem. Well I found a way to check if connection is present from Python process, however we still need to close the connection somewhere on java side right? If destroy is not called, how will this happen?

The connection will drop itself if we just kill the process so it will work fine for this case. And I have actually tested it. But what about Spark scenario? Unless the whole process running the job dies connection won't get closed (from java side).

(if we don't figure this out I think for streaming the only thing we can do is to start the process during every transform)

ps. regarding the authentication work, there is also a quick question here in comments: https://wiki.cask.co/display/CE/Python+transform+improvements hope you can look as well.

File: docs/PythonEvaluator-transform.md (hunk 3 above)
Comment (albertshau) [reply docs/PythonEvaluator-transform.md]: I see. Then maybe something like:

This value is only used in native mode.

'Works only' makes it sound like it's broken in other modes.

File: docs/PythonEvaluator-transform.md [hunk 14]
```diff
@@ -21,18 +21,20 @@ For example:
...
+It also requires that the py4j library is available on every node.
+To install py4j with Python 2, run ``pip install py4j``. To install py4j with Python 3, run ``pip3 install py4j``.
 
-**pythonBinary:** Path to binary which will run the python code. E.g. ``/usr/bin/python3``.
+**Python Binary:** Path to binary which will run the python code. E.g. ``/usr/bin/python3``.
 Works only in native mode.
 
-**pythonPath:** PYTHONPATH environment variable. Allows to include libs from various locations. Path are separated with a colon.
+**PYTHONPATH:** PYTHONPATH environment variable. Allows to include libs from various locations.
 Works only in native mode.
```
Comment (albertshau) [inline docs/PythonEvaluator-transform.md]: same thing here

File: widgets/PythonEvaluator-transform.json [hunk 15]
```diff
@@ -14,6 +14,33 @@
...
+      "properties": [
+        {
+          "widget-type": "textbox",
+          "label": "Python Binary",
+          "name": "pythonBinary"
+        },
+        {
+          "widget-type": "textbox",
+          "label": "PYTHONPATH",
+          "name": "pythonPath"
```
Comment (albertshau) [inline widgets/PythonEvaluator-transform.json:46]: Let's add a default value here of 'python'.

Even though there is a default in the code, it is still nice from a user perspective to have the default filled in the UI. Our newer plugins follow this pattern where backend default values are also specified in the UI widget.

File: pom.xml (hunk 6 above)
Comment (albertshau) [reply pom.xml:263]: does this mean the code doesn't compile yet without those changes? Can you open a PR for those?

Comment (aonischuk) [reply pom.xml:263]: that is correct. Will open a PR for that.

**Comment by aonischuk** [comment]
@albertshau I have addressed your new review comments. Can you please check this again? :)

**Comment by albertshau** [comment]
In normal scenarios, destroy() will be called in mapreduce pipelines. If the process was killed some other way, it's the same to the python process as the connection will no longer be there since the jvm has shut down. In Spark it's the same, the process will shutdown and end, it's just that destroy() won't be called. So we're not as worried about a java process hanging around never calling destroy(), it's more that they have ended without calling it.

**Comment by albertshau** [comment]
these changes lgtm. Can you squash them into a single commit? I'll merge after that.

**Comment by aonischuk** [comment]
@albertshau 

> In Spark it's the same, the process will shutdown and end, it's just that destroy() won't be called. So we're not as worried about a java process hanging around never calling destroy(), it's more that they have ended without calling it.

I actually tried it on standalone cdap with Spark Streaming. And the java server does not die after streaming job finishes. Are you sure it's started in a separate process? I didn't see a new process and it's strange why java server didn't die.

**Comment by aonischuk** [comment]
@albertshau  I have squashed all the changes into one commit. Please feel free to merge.

**Comment by albertshau** [comment]
Hm I guess the sandbox has this issue because everything runs in one JVM. But that would mean the gateway is still running in some thread somewhere? In distributed mode it will be a separate container.

In any case, I'm not sure we can do much from the plugin side for that. Fixing it would have to be a CDAP/pipeline level change.

**Pull request merged**